Add reduce keyword to NLLLoss and NLLLoss2d

zou3519 opened the pull request:
### Summary
Adding a reduce keyword to NLLLoss as per https://github.com/pytorch/pytorch/issues/264. 
By default reduce is True and gives the usual behavior; when reduce is False the loss returns
a loss per element.

### Test Plan
Added a few new module unit tests. Each unit test is analogous to a criterion test (in the case that reduce is True):
- NLLLoss, no reduce
- NLLLoss, no reduce, ignore index
- NLLLoss, no reduce, weights
- NLLLoss, no reduce, weights and ignore index
- NLLLoss, no reduce, negative ignore index
- NLLLoss2d, no reduce
- NLLLoss2d, weights
- NLLLoss2d, ignore index

Also ran gradcheck by hand for several of the above cases. Example code:
```
import torch
import torch.autograd as autograd
import torch.nn as nn

loss = nn.NLLLoss(size_average=False)
loss2 = nn.NLLLoss(reduce=False)
input = autograd.Variable(torch.randn(3, 5).double().cuda(), requires_grad=True)
target = autograd.Variable(torch.LongTensor([1, 0, 4]).cuda(), requires_grad=False)
autograd.gradcheck(loss2, (input, target), raise_exception=True)
sum(loss2(input, target)) - loss(input, target)  # check that this is 0
```

zou3519 committed:
API changes

zou3519 committed:
Implement reduce for THNN ClassNLLCriterion

zou3519 committed:
Implement reduce keyword for THCUNN ClassNLLCriterion

zou3519 committed:
Implement reduce for THNN SpatialClassNLLCriterion

zou3519 committed:
Implement reduce for THCUNN SpatialClassNLLCriterion

zou3519 committed:
Make legacy NLLLoss work

zou3519 committed:
Docs for NLLLoss reduce

zou3519 committed:
reduce keyword for double backwards NLLLoss

zou3519 committed:
reduce=False tests

colesbury commented on torch/lib/THCUNN/generic/ClassNLLCriterion.cu:
```diff
@@ -46,6 +43,37 @@ void THNN_(ClassNLLCriterion_updateOutput)(
 
+  if (!reduce && n_dims == 2) {
+    THCTensor_(resize1d)(state, output, batch_size);
+    bool not_contiguous_weights = weights && !THCTensor_(isContiguous)(state, weights);
```
Use:

```
weights = THCTensor_(newContiguous)(state, weights)
...
THCTensor_(free)(state, weights)
```

i.e. don't guard it with a contiguous check. `newContiguous` will already avoid the copy and just incref the tensor if it's already contiguous.

colesbury commented on torch/legacy/nn/ClassNLLCriterion.py:
```diff
@@ -27,24 +27,28 @@ def updateOutput(self, input, target):
     def updateGradInput(self, input, target):
         self.gradInput.resize_as_(input).zero_()
         target = target.long()
+        implicit_gradOutput = torch.ones(1).type_as(input)
```
torch.ones_like(input)

zou3519 replied:
The implicit gradOutput should be a tensor with one element, the value 1 inside it, so I don't think `torch.ones_line(input)` is what I'm looking for here. Is there a nicer way to write this?

colesbury replied:
My mistake -- this is fine

apaszke replied:
Actually I think this won't work for CUDA

zou3519 replied:
How so? Do you mean that if input is CUDA then implicit_gradOutput won't be?

apaszke replied:
Um, sorry it should work. However I just realized that it might not respect the device of input (which is a bug in type_as)

colesbury commented on torch/lib/THCUNN/generic/ClassNLLCriterion.cu:
```diff
@@ -97,15 +125,16 @@ void THNN_(ClassNLLCriterion_updateGradInput)(
            THCState *state,
            THCTensor *input,
            THCIndexTensor *target,
+           THCTensor *gradOutput,
```
If you're going to take in a `gradOutput` you need to make sure you use it in both the reduce and non-reduce cases.

colesbury commented on torch/lib/THCUNN/generic/SpatialClassNLLCriterion.cu:
```diff
@@ -29,6 +29,24 @@ void THNN_(SpatialClassNLLCriterion_shapeCheck)(
   }
 }
 
+void THNN_(SpatialClassNLLCriterion_gradOutput_no_reduce_shapeCheck)(
```
`static void ...`

colesbury commented on torch/lib/THCUNN/generic/SpatialClassNLLCriterion.cu:
```diff
@@ -148,4 +231,6 @@ void THNN_(SpatialClassNLLCriterion_updateGradInput)(
   THCTensor_(free)(state, input);
 }
 
+#undef TO_DEVICE
```
What's this?

zou3519 replied:
Forgot to remove that, thanks for pointing it out!

zou3519 committed:
Addressed comments

zou3519 committed:
Fix trailing whitespace

zou3519 committed:
Fix test failures in legacy nn

ezyang commented:
@colesbury, @apaszke, can we take this now?

colesbury commented on torch/lib/THCUNN/generic/ClassNLLCriterion.cu:
```diff
@@ -135,9 +163,38 @@ void THNN_(ClassNLLCriterion_updateGradInput)(
   target = THCIndexTensor_(newContiguous)(state, target);
 
+  THCUNN_check_dim_size(state, gradOutput, 1, 0, 1);
+  real gradOutput_data = THCTensor_(get1d)(state, gradOutput, 0);
```
This will cause a host-device synchronization. I think it's probably better to pass the data pointer into the kernel and use it there.

(Same with the other `THCTensor_(get1d)` call below)

colesbury reviewed (commented):
This looks good. I'd like to get the ATen/onnx stuff in first and then rebase this on top of it. I can help with any rebasing

apaszke commented on test/test_nn.py:
```diff
@@ -3561,8 +3561,134 @@ def mseloss_no_reduce_module_test():
         reference_fn=lambda i, m: (i - target).pow(2))
 
 
+class TestNLLLoss(torch.nn.modules.module.Module):
```
Why do we have to add these test modules, plus extra functions that return dicts, which seem to be valid literals that would be specified in the list itself? I don't understand what's the value here

zou3519 replied:
I made a module that wraps NLLLoss so that its forward function takes one argument, the input. The [module tests](https://github.com/pytorch/pytorch/blob/d5604aea0bf23c47a915f470c903beb3b6915090/test/common_nn.py#L646) assume that the module's forward function takes one argument and I wanted to reuse the module tests because the criterion tests assume the output of the loss is scalar.

Good point about the return dicts. I think that once I add in reference functions for each of these, I'll need the return dict because a target should be passed to the test module constructor and to the reference function, just like [this](https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L3639)

Alternatively to make testing cleaner I could write a new sort of criterion test logic that doesn't assume the output of the loss is scalar.

apaszke replied:
You could always use the newly added `wrap_functional` to test a partially applied `NLLLoss` with different arguments, including the targets.

zou3519 replied:
Thanks! That was useful

apaszke commented on torch/lib/THNN/generic/ClassNLLCriterion.c:
```diff
@@ -30,6 +29,32 @@ void THNN_(ClassNLLCriterion_updateOutput)(
+    int i;
+    #pragma omp parallel for private(i)
+    for (i = 0; i < batch_size; i++) {
+      int cur_target = THIndexTensor_(get1d)(target, i);
```
Never use `get1d` in tight loops. It does bound checking

apaszke commented on torch/lib/THNN/generic/ClassNLLCriterion.c:
```diff
@@ -30,6 +29,32 @@ void THNN_(ClassNLLCriterion_updateOutput)(
+    for (i = 0; i < batch_size; i++) {
+      int cur_target = THIndexTensor_(get1d)(target, i);
+      if (cur_target == ignore_index) {
+        THTensor_(set1d)(output, i, 0.0f);
```
Same for `set1d`

apaszke commented on torch/lib/THNN/generic/ClassNLLCriterion.c:
```diff
@@ -115,20 +135,54 @@ void THNN_(ClassNLLCriterion_updateGradInput)(
+        continue;
+      }
+      real weight = weights ? THTensor_(get1d)(weights, cur_target) : 1.0f;
+      THTensor_(set2d)(gradInput, i, cur_target, -weight * THTensor_(get1d)(gradOutput, i));
```
As above

apaszke commented on torch/lib/THNN/generic/ClassNLLCriterion.c:
```diff
@@ -115,20 +135,54 @@ void THNN_(ClassNLLCriterion_updateGradInput)(
+
+  real *total_weight_data = THTensor_(data)(total_weight);
+
+  if (!(*total_weight_data > 0)) {
```
It seems that `total_weight_data <= 0` would be clearer. Also, what's the purpose of this check?

zou3519 replied:
Agreed. I'm not really sure what the purpose of this check is (it was in the [non-reduce code](https://github.com/pytorch/pytorch/blob/d5604aea0bf23c47a915f470c903beb3b6915090/torch/lib/THNN/generic/ClassNLLCriterion.c#L105) before. I'll investigate a little and see if it's necessary.

apaszke commented on torch/lib/THNN/generic/ClassNLLCriterion.c:
```diff
@@ -115,20 +135,54 @@ void THNN_(ClassNLLCriterion_updateGradInput)(
+    return;
+  }
+
+  ignore_index -= TH_INDEX_BASE;
```
Non-reduce kernel will not see this change

apaszke commented on torch/lib/THNN/generic/SpatialClassNLLCriterion.c:
```diff
@@ -84,15 +129,45 @@ void THNN_(SpatialClassNLLCriterion_updateGradInput)(
           THTensor *weights,
           THTensor *total_weight,
-          int64_t ignore_index)
+          int64_t ignore_index,
```
It seems that we don't include `TH_INDEX_BASE` in the offset here

zou3519 replied:
TH_INDEX_BASE = 0. I think it's used because everything in lua torch is one-indexed? Should I remove TH_INDEX_BASE entirely from this file, or assume that it could potentially equal 1?

apaszke commented on torch/lib/THNN/generic/SpatialClassNLLCriterion.c:
```diff
@@ -121,8 +196,10 @@ void THNN_(SpatialClassNLLCriterion_updateGradInput)(
+      int index = b * sample_size + cur_target * map_size + elem;
+      gradInput_data[index] =
         -(weights ? weights_data[cur_target] : 1.0f) / normalize;
+      gradInput_data[index] *= THTensor_(get1d)(gradOutput, 0);
```
can't you add this multiplication to previous assignment?

apaszke commented on torch/lib/THCUNN/ClassNLLCriterion.cu:
```diff
@@ -34,6 +37,47 @@ __global__ void cunn_ClassNLLCriterion_updateOutput_kernel1(Dtype *output,
+    int64_t ignore_index) {
+
+  CUDA_KERNEL_LOOP(index, batch_size) {
+    int64_t cur_target = target[index];
```
You should avoid using `int64_t` math in CUDA kernels. It's way more expensive than 32-bit math. In most cases we instantiate two kernels using two kinds of indexing, based on tensor sizes

apaszke reviewed (commented):
Haven't reviewed the actual kernels/math. Let's make sure we have `reference_fn`s for some of these tests before merging. Just found a few worrying things.

zou3519 committed:
Replaced slow get/set fns. Don't use int64_t in kernels.

ezyang reviewed (dismissed):
This needs a rebase at least (ping @colesbury if you need help) and it looks like there are some comments from Adam which still need addresing.

zou3519 committed:
Add reference functions for all NLLLoss and NLLLoss2d test cases

zou3519 committed:
Rebase: add reduce keyword to aten declarations of NLLLoss

zou3519 committed:
Use TH_INDEX_BASE in NLLLoss for consistency

zou3519 commented:
Rebased and addressed comments. Main changes were: 
- rebasing to work with the new aten autograd
- replacing slow get/set tensor functions with faster equivalents
- supporting TH_INDEX_BASE for consistency

zou3519 committed:
Fix legacy ClassNLLCriterion tests
